File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    
    private static long findEmptyRoom(Map<Long, Long> reserved, long current) {
        if (!reserved.containsKey(current)) {
            reserved.put(current, current + 1);
            return current;
        }
        
        long next = findEmptyRoom(reserved, reserved.get(current));
        reserved.put(current, next);
        return next;
    }
    
    public long[] solution(long k, long[] room_number) {
        long[] answer = new long[room_number.length];
        Map<Long, Long> reserved = new HashMap<>();
        
        for (int i = 0; i < room_number.length; i++) {
            answer[i] = findEmptyRoom(reserved, room_number[i]);
        }
        
        return answer;
    }
}

File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.py
import sys
sys.setrecursionlimit(2_000)


def find_empty_room(rooms, x):
    if x not in rooms:
        rooms[x] = x + 1
        return x
    
    empty_room = find_empty_room(rooms, rooms[x])
    rooms[x] = empty_room + 1
    return empty_room
    

def solution(k, room_number):
    rooms = dict()
    return [find_empty_room(rooms, x) for x in room_number]
